File: src/market-tracker-android/app/src/main/java/pt/isel/markettracker/domain/model/market/price/Price.kt
package pt.isel.markettracker.domain.model.market.price

import java.time.LocalDateTime

data class Price(
    val regularPrice: Int,
    val finalPrice: Int,
    val promotion: Promotion?,
    val createdAt: LocalDateTime
)

data class Promotion(
    val percentage: Int,
    val createdAt: LocalDateTime
)

File: src/market-tracker-webapi/Application/Domain/Schemas/Market/Retail/Sales/Pricing/Price.cs
using market_tracker_webapi.Application.Utils;

namespace market_tracker_webapi.Application.Domain.Schemas.Market.Retail.Sales.Pricing;

public record Price(int BasePrice, int FinalPrice, Promotion? Promotion, DateTime CreatedAt)
{
    public Price(int basePrice, Promotion? promotion, DateTime createdAt) : this(
        basePrice,
        basePrice.ApplyPercentage(promotion?.Percentage),
        promotion,
        createdAt
    )
    {
    }
}

public record PriceId(
    string Value
);